File: link_list/link_list.c
// Copyright 2022 Zncl2222
#include "link_list.h"

void create_link_list(struct LinkList **L, int L_size, datatype init_value,
                    bool random_value) {
    struct LinkList *temp, *rear;
    srand(time(0));

    *L = (struct LinkList *)malloc(sizeof(LinkNode));
    rear = *L;
    for (int i = 0; i < L_size + 1; i++) {
        temp = (struct LinkList *)malloc(sizeof(LinkNode));

        if (random_value == true)
            temp->val = rand() % 10 + 1; // NOLINT
        else
            temp->val = init_value;

        rear->next = temp;
        rear = temp;
    }

    rear->next = NULL;
}

void delete_link_list(struct LinkList **L) {
    struct LinkList *a, *b;
    a = *L;

    while (a != NULL) {
        b = a->next;
        free(a);
        a = b;
    }
    *L = NULL;
}

void insert_link_node(struct LinkList **L, int target_idx,
                    datatype target_value) {
    struct LinkList *a, *b;
    int curr_idx = 1;

    a = *L;

    while (a != NULL && curr_idx < target_idx + 1) {
        a = a->next;
        curr_idx++;
    }

    if (a == NULL || curr_idx > target_idx + 1) {
        printf("ERROR: Out of index\n");
        return;
    } else {
        b = (struct LinkList *)malloc(sizeof(LinkNode));
        b->val = target_value;
        b->next = a->next;
        a->next = b;
    }
}

void delete_link_node(struct LinkList **L, int target_idx) {
    struct LinkList *a, *b;
    int curr_idx = 1;

    a = *L;

    while (a->next != NULL && curr_idx < target_idx + 1) {
        a = a->next;
        curr_idx++;
    }

    if (a->next == NULL || curr_idx > target_idx + 1) {
        printf("ERROR: Target is not exist\n");
    } else {
        b = a->next;
        a->next = b->next;
        free(b);
    }
}

int link_list_length(struct LinkList **L) {
    struct LinkList *temp;
    int count = 0;

    temp = *L;

    while (temp->next != NULL) {
        temp = temp->next;
        count++;
    }

    return count;
}

datatype get_element_val(struct LinkList **L, datatype target_idx) {
    struct LinkList *temp;
    int target_value;

    temp = *L;

    for (int i = 0; i < target_idx; i++) {
        temp = temp->next;
    }

    target_value = temp->val;

    return target_value;
}

datatype get_element_idx(struct LinkList **L, datatype target_value) {
    struct LinkList *temp;
    int target_idx = 0;

    temp = (*L)->next;

    while (temp != NULL) {
        if (temp->val == target_value) {
            return target_idx;
        }
        temp = temp->next;
        target_idx++;
    }

    return -1;
}

void traverse(struct LinkList **L) {
    struct LinkList *temp;
    int Len;
    temp = (*L)->next;

    Len = link_list_length(L);

    printf("DataValue=\n");
    for (int i = 0; i < Len; i++) {
        printf("%d ", temp->val);
        temp = temp->next;
    }
    printf("\n");
}

File: link_list/link_list.py
class LinkNode:
    def __init__(self, val):
        self.val = val
        self.next = None


def create_link_list(nums: list):
    if len(nums) == 0:
        return None
    head = LinkNode(nums[0])
    cur = head
    for i in range(1, len(nums)):
        cur.next = LinkNode(nums[i])
        cur = cur.next
    return head


def delete_link_list(head: LinkNode):
    while head:
        head = head.next
        del head


def print_link_list(head: LinkNode):
    while head:
        if head.next is None:
            print(head.val)
            break
        print(head.val, end=", ")
        head = head.next


def insert_link_node(head: LinkNode, target_index: int, val: int):
    if head is None:
        return LinkNode(val)
    cur = head
    for _ in range(target_index - 1):
        cur = cur.next
    cur.next = LinkNode(val)
    return head


def get_link_list_length(head: LinkNode):
    length = 0
    while head:
        length += 1
        head = head.next
    return length


def get_element_val(head: LinkNode, index: int):
    if index < 0:
        return None
    cur = head
    for _ in range(index):
        cur = cur.next
    return cur.val


def get_element_idx(head: LinkNode, val: int):
    if head is None:
        return -1
    cur = head
    idx = 0
    while cur:
        if cur.val == val:
            return idx
        cur = cur.next
        idx += 1
    return -1
